library/data/recipe/src/main/java/org/quiltmc/qsl/recipe/api/builder/VanillaRecipeBuilders.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright 2022 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.recipe.api.builder;
+
+
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+
+import it.unimi.dsi.fastutil.chars.Char2ObjectMap;
+import it.unimi.dsi.fastutil.chars.Char2ObjectOpenHashMap;
+import it.unimi.dsi.fastutil.chars.CharArraySet;
+
+import net.minecraft.item.Item;
+import net.minecraft.item.ItemConvertible;
+import net.minecraft.item.ItemStack;
+import net.minecraft.recipe.BlastingRecipe;
+import net.minecraft.recipe.CampfireCookingRecipe;
+import net.minecraft.recipe.Ingredient;
+import net.minecraft.recipe.Recipe;
+import net.minecraft.recipe.ShapedRecipe;
+import net.minecraft.recipe.ShapelessRecipe;
+import net.minecraft.recipe.SmeltingRecipe;
+import net.minecraft.recipe.SmokingRecipe;
+import net.minecraft.recipe.StonecuttingRecipe;
+import net.minecraft.tag.TagKey;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.collection.DefaultedList;
+
+import org.quiltmc.qsl.recipe.impl.VanillaRecipeBuildersImpl;
+
+/**
+ * Provides some recipe builders for Vanilla recipes.
+ */
+public final class VanillaRecipeBuilders {
+	private VanillaRecipeBuilders() {
+		throw new UnsupportedOperationException("VanillaRecipeBuilders only contains static definitions.");
+	}
+
+	/**
+	 * Returns the list of ingredients for shaped crafting recipes.
+	 *
+	 * @param pattern the pattern of the shaped crafting recipe
+	 * @param keys    the keys and ingredients of the recipe
+	 * @param width   the width of the shaped crafting recipe
+	 * @param height  the height of the shaped crafting recipe
+	 * @return the ingredients
+	 * @throws IllegalStateException if a key has no assigned ingredient or if there is an ingredient but no assigned key
+	 */
+	public static DefaultedList<Ingredient> getIngredients(String[] pattern, Char2ObjectMap<Ingredient> keys, int width, int height) {
```

Comment:
Is fastutil in public api really necessary here?

Reply:
Is it really that bad?

Given that the type exists already, and Project Valhalla still far, it kinds of feel wrong to use a Map.

Reply:
it's technically poor practice, but eh--it's fine